src/binning_based/rectangular/TransferOperator.jl

```diff
@@ -0,0 +1,383 @@
+using GroupSlices, DelayEmbeddings, SparseArrays
+
+export TransferOperator, 
+    transferoperator, TransferOperatorApproximationRectangular,
+    InvariantMeasureEstimate, invariantmeasure,
+    binhist
+
+"""
+    TransferOperator(r::RectangularBinning)
+
+A probability estimator based on binning data into rectangular boxes dictated by 
+the binning scheme `r`, then estimating the transfer (Perron-Frobenius) operator 
+over the bins. Assumes that the input data are sequential.
+
+The transfer operator ``P^{N}``is computed as an `N`-by-`N` matrix of transition 
+probabilities between the states defined by the partition elements, where `N` is the 
+number of boxes in the partition that is visited by the orbit/points. The left invariant
+distribution ``\\rho^{N}``, where ``\\rho^{N}P^{N} = \\rho^{N}``, approximates the 
+invariant density of the system subject to the partition `r`. Thus, ``\\rho^{N}`` can 
+be taken as a probability distribution over the partition elements.
+
+This implementation follows the approach in Diego et al. (2019)[^Diego2019].
+
+See also: [`RectangularBinning`](@ref).
+
+[^Diego2019]: Diego, D., Haaga, K. A., & Hannisdal, B. (2019). Transfer entropy computation using the Perron-Frobenius operator. Physical Review E, 99(4), 042212.
+"""
+struct TransferOperator{R} <: BinningProbabilitiesEstimator
+    binning::R
+    
+    function TransferOperator(binning::R) where R <: RectangularBinning
+        new{R}(binning)
+    end
+end
+
+# If x is not sorted, we need to look at all pairwise comparisons
+function inds_in_terms_of_unique(x)
+    U = unique(x)
+    N = length(x)
+    Nu = length(U)
+    inds = zeros(Int, N)
+    
+    for j = 1:N
+        xⱼ = view(x, j)
+        for i = 1:Nu
+            # using views doesn't allocate
+            @inbounds if xⱼ == view(U, i)
+                inds[j] = i
+            end
+        end
+    end
+    
+    return inds
+end
+
+# Taking advantage of the fact that x is sorted reduces runtime by 1.5 orders of magnitude 
+# for datasets of >100 000+ points
+function inds_in_terms_of_unique_sorted(x) # assumes sorted
+    @assert issorted(x)
+    U = unique(x)
+    N, Nu = length(x), length(U)
+    prev = view(x, 1)
+    inds = zeros(Int, N)
+    uidx = 1
+    @inbounds for j = 1:N
+        xⱼ = view(x, j)
+        # if the current value has changed, then we know that the corresponding index
+        # for the unique point must be incremented by 1
+        if xⱼ != prev
+            prev = xⱼ
+            uidx += 1
+        end
+        inds[j] = uidx
+    end
+    
+    return inds
+end
+
+function inds_in_terms_of_unique(x, sorted::Bool)
+    if sorted 
+        return inds_in_terms_of_unique_sorted(x)
+    else
+        return inds_in_terms_of_unique(x)
+    end
+end
+
+inds_in_terms_of_unique(x::Dataset) = inds_in_terms_of_unique(x.data)
+
+"""
+    TransferOperatorApproximationRectangular(to, ϵ::RectangularBinning, mini, edgelengths, bins)
```

Comment:
Does this need to be exported and added in the docs? My guess is you need it in some package downstream?

Reply:
EDIT: yes, all later functions return and use this type.

Reply:
Yes, this is used in CausalityTools later on too.